crates/driver/src/domain/competition/solution/encoding.rs

```diff
@@ -145,6 +155,9 @@ pub fn tx(
         clearing_prices.push(price.sell_price);
         clearing_prices.push(price.buy_price);
 
+        trade.sell_token_index = (tokens.len() - 2).into();
+        trade.buy_token_index = (tokens.len() - 1).into();
```

Comment:
Does it make sense here to avoid these mutations by first retrieving the price, pushing to the tokens vec and only then creating a trade with the specified indices?

Reply:
The initial code had the `tokens.push()` right before building the trade which meant the trade had the correct token indices on initialization. IMO having the token indices initialised at the same place where we modify the `tokens` array kind of shows the dependency between trade and token data neatly grouped into 1 chunk of code.

Reply:
Another way to go about it would be to have a new type for token indices. That way pushing tokens into the array would happen before preparing and pushing trades and prices into the settlement (which I dislike) but the dependency would be enforced by the type system so you can't do it wrong.

Specifically I mean sth like:
```rust
struct TokenIndex(usize);

struct Tokens(Vec<H160>):

impl Tokens {
    fn add_token(&mut self, token: H160) -> TokenIndex {
        self.push(token);
        TokenIndex(self.len() - 1)
    }
}
```

Then `Trade` would have a `TokenIndex` instead of `eth::U256` which means you first have to:
```rust
let sell_token_index = tokens.push(price.sell_token);
let buy_token_index = tokens.push(price.buy_token);

// trade can later only be built using these new token indices.
```